Add support, tests for tracing in Clojurescript

alex-dixon opened the pull request:
- Received warnings about "No such var" in cljs.test namespace before making any changes. Fixed by requiring cljs.test. 
- Made tracing CLJC
- Rewrote tests in test_tracing using the (awesome) def-rules-test macro

I'm receiving the following errors on master when running `lein test` before making other changes. These errors happen for me in this branch and did in #311 as well:
```
ERROR in (test-extract-test-inspect) (core.clj:1524)
expected: (not (and (symbol? term) (.startsWith (name term) "?__gen")))
  actual: 
                      clojure.main.main         main.java:   37
                                    ...                        
                      clojure.main/main          main.clj:  421
                  clojure.main/null-opt          main.clj:  343
                clojure.main/initialize          main.clj:  308
                  clojure.main/init-opt          main.clj:  280
               clojure.main/load-script          main.clj:  275
                                    ...                        
                          user/eval2848         REPL Input     
                       user/eval2848/fn         REPL Input     
                    user/eval2848/fn/fn         REPL Input      (repeats 2 times)
                     clojure.core/apply          core.clj:  630
                                    ...                        
                 clojure.test/run-tests          test.clj:  762
                     clojure.core/apply          core.clj:  632
                                    ...                        
                clojure.core/merge-with          core.clj: 2936
                   clojure.core/reduce1          core.clj:  900
                   clojure.core/reduce1          core.clj:  909
                      clojure.core/next          core.clj:   64
                                    ...                        
                    clojure.core/map/fn          core.clj: 2624
                   clojure.test/test-ns          test.clj:  747
             clojure.test/test-all-vars          test.clj:  728
                 clojure.test/test-vars          test.clj:  718
       clojure.test/compose-fixtures/fn          test.clj:  681
           clojure.test/default-fixture          test.clj:  674
    clojure.test/compose-fixtures/fn/fn          test.clj:  681
           schema.test/validate-schemas          test.clj:   11
                                    ...                        
   schema.test/validate-schemas/fn10620          test.clj:   11
schema.test/validate-schemas/fn10620/fn          test.clj:   11
              clojure.test/test-vars/fn          test.clj:  722
           clojure.test/default-fixture          test.clj:  674
           clojure.test/test-vars/fn/fn          test.clj:  722
                  clojure.test/test-var          test.clj:  704
               clojure.test/test-var/fn          test.clj:  704
            clara.tools.test-inspect/fn  test_inspect.clj:  182
         clara.tools.test-inspect/fn/fn  test_inspect.clj:  183
                      clojure.core/name          core.clj: 1524
java.lang.ClassCastException: java.lang.Long cannot be cast to clojure.lang.Named

ERROR in (test-extract-test-inspect) (core.clj:1524)
expected: (not (and (symbol? term) (.startsWith (name term) "?__gen")))
  actual: 
                      clojure.main.main         main.java:   37
                                    ...                        
                      clojure.main/main          main.clj:  421
                  clojure.main/null-opt          main.clj:  343
                clojure.main/initialize          main.clj:  308
                  clojure.main/init-opt          main.clj:  280
               clojure.main/load-script          main.clj:  275
                                    ...                        
                          user/eval2848         REPL Input     
                       user/eval2848/fn         REPL Input     
                    user/eval2848/fn/fn         REPL Input      (repeats 2 times)
                     clojure.core/apply          core.clj:  630
                                    ...                        
                 clojure.test/run-tests          test.clj:  762
                     clojure.core/apply          core.clj:  632
                                    ...                        
                clojure.core/merge-with          core.clj: 2936
                   clojure.core/reduce1          core.clj:  900
                   clojure.core/reduce1          core.clj:  909
                      clojure.core/next          core.clj:   64
                                    ...                        
                    clojure.core/map/fn          core.clj: 2624
                   clojure.test/test-ns          test.clj:  747
             clojure.test/test-all-vars          test.clj:  728
                 clojure.test/test-vars          test.clj:  718
       clojure.test/compose-fixtures/fn          test.clj:  681
           clojure.test/default-fixture          test.clj:  674
    clojure.test/compose-fixtures/fn/fn          test.clj:  681
           schema.test/validate-schemas          test.clj:   11
                                    ...                        
   schema.test/validate-schemas/fn10620          test.clj:   11
schema.test/validate-schemas/fn10620/fn          test.clj:   11
              clojure.test/test-vars/fn          test.clj:  722
           clojure.test/default-fixture          test.clj:  674
           clojure.test/test-vars/fn/fn          test.clj:  722
                  clojure.test/test-var          test.clj:  704
               clojure.test/test-var/fn          test.clj:  704
            clara.tools.test-inspect/fn  test_inspect.clj:  182
         clara.tools.test-inspect/fn/fn  test_inspect.clj:  183
                      clojure.core/name          core.clj: 1524
java.lang.ClassCastException: java.lang.Long cannot be cast to clojure.lang.Named
```

WilliamParker commented:
I think [this line](https://github.com/cerner/clara-rules/blob/0.15.0/src/main/clojure/clara/tools/tracing.clj#L12) needs to be changed, probably to use [ex-info](https://clojuredocs.org/clojure.core/ex-info), since the use of [IllegalStateException](https://docs.oracle.com/javase/8/docs/api/java/lang/IllegalStateException.html) relies on Java interop.  Strictly speaking, this is an API change, but it is very minor so I think it is fine although we can note it in the changelog.

WilliamParker commented on src/test/clojure/clara/tools/test_tracing.cljc:
```diff
@@ -0,0 +1,147 @@
+            :accum-reduced :left-retract :left-activate :add-facts
+            :alpha-activate :right-activate :accum-reduced]
+
+           (map :type (t/get-trace session))))))
```
Minor: It would be better to have a one line of whitespace between tests.

alex-dixon replied:
Struggling with formatting in general. I use Parinfer and IntelliJ and it wants to change the formatting in Clara's files in ways that break the code. Do you have editor settings that you share?

WilliamParker replied:
No, I think it just hasn't come up much.  It would perhaps be ideal to have some sort of editor-independent automatic formatter e.g. a Leiningen formatting plugin, but I haven't investigated that in any meaningful way.

WilliamParker commented:
Which var were the "Received warnings about "No such var" in cljs.test namespace before making any changes. Fixed by requiring cljs.test." warnings for?  I don't think adding the cljs.test require is a problem if we need it for some reason, but it isn't obvious to me why we need it since the things we're bringing in are macros using :requires-macros. I also don't see such warnings in the Travis build.

WilliamParker commented:
I think we'll want to have the commits changing tracing.clj to .cljc and the commit for uplifting the tests to def-rules-test squashed together since they are logically one change (an enhancement and testing of that enhancement).  IMO adding the cljs.test :require is an ambiguous case, since it is an unrelated change, albeit a very minor one.  I'd be fine with that one being either its own commit or squashed together for my part.

WilliamParker commented:
On the error: That is strange.  The error almost looks like and isn't short-circuiting in 

````
(not (and (symbol? term)
          (.startsWith (name term) "?__gen" )))
````
but a bug like that would be very surprising to me.  Do you have a way to reproduce this on another machine?  Are you doing anything in your Leiningen setup that could impact this?  Since "is" is a macro development tooling can modify its behavior, including its code body; I've seen special-casing of forms like (is (= a b)) before in development tooling.

WilliamParker commented:
This isn't actually running the tests in ClojureScript since this test namespace wasn't added to the [CLJS test runner namespace.  You can search the logs of the Travis CI build to verify this.

alex-dixon commented:
> Which var were the "Received warnings about "No such var" in cljs.test namespace before making any changes. Fixed by requiring cljs.test." warnings for?
Wish I could tell you. I removed the require for cljs.test and am no longer receiving the warnings. I think they were for `is`, `testing`...whatever macros we were using. Could have been caused by a typo or something. Sorry if I've created any confusion. 😅 
> This isn't actually running the tests in ClojureScript
But...the tests are passing! 😄 Fixed.
> Are you doing anything in your Leiningen setup that could impact this?
Wow. Yes. I commented out `[venantius/ultra "0.5.1" :exclusions [org.clojure/clojure]]` in my profiles.clj and the error went away. In addition, the CLJS tests are actually running now. I guess I was only running CLJ before but now I see:
```
...clj tests
Ran 412 tests containing 1489 assertions.
0 failures, 0 errors.
Running ClojureScript test: phantom-simple

Testing clara.test-rules

Testing clara.test-common

Testing clara.test-salience

Testing clara.test-testing-utils

Testing clara.test-complex-negation

Testing clara.test-accumulators

Testing clara.tools.test-tracing

Ran 39 tests containing 97 assertions.
0 failures, 0 errors.
Success!
Running ClojureScript test: phantom-advanced

Testing clara.test-rules

Testing clara.test-common

Testing clara.test-salience

Testing clara.test-testing-utils

Testing clara.test-complex-negation

Testing clara.test-accumulators

Testing clara.tools.test-tracing

Ran 39 tests containing 97 assertions.
0 failures, 0 errors.
Success!
```

WilliamParker commented on src/main/clojure/clara/tools/tracing.cljc:
```diff
@@ -118,4 +118,4 @@
                          (first))]
     (.-trace ^PersistentTracingListener listener)
-    (throw (IllegalStateException. "No tracing listener attached to session."))))
+    (throw (ex-info "No tracing listener attached to session." {}))))
```
Minor: it would be better to have the session in the data map.  i.e. 

````
(throw (ex-info "No tracing listener attached to session." {:session session}))
````
It is nice to have the offending object in the exception so that it can be examined in the REPL when an exception is thrown during development.

WilliamParker commented:
Regarding the history, my preference would be to have all this squashed into one commit.

I made one minor comment at https://github.com/cerner/clara-rules/pull/315/files#r124424540
Otherwise the changes look good to me.
